### ISSUE TITLE
Page Encryption

### ISSUE THREAD

**Issue by cdelahousse** [SteamedPears/Code-Review#21]
_Imported issue by **[Gankro](http://github.com/Gankro)** from 2012-07-23T02:59:24Z_
_Originally opened as https://github.com/SteamedPears/CodeReviewClientMaster/issues/71_

---

Implement page encryption using new form fields.

If "encrypt page" is checked of, encrypt the code client side symmetrically using salt+password. Also store a hash of salt+password and the salt itself. 

If someone navigates to the code the server will send the salt and request authentication. If the client returns hashed salt+password (as stored on server), server sends encrypted code, which is decrypted client side using the user's entered password. If the hash is incorrect have the server wait 5 seconds before replying. After several failed attempts IP blocking may be necessary.

Comments should also be encrypted and decrypted in the same way as code (though if you can request/submit comments you've already been authenticated). 

For hashing I recommend BCrypt. For encryption AES should be sufficient. BCrypt should also be able to handle salt generation.

**Comment by spratt** [comment]
This security model may be broken, since the user has to trust we aren't storing their password.  Additionally, unless we use SSL, we can't assume the code will get to the user unmodified by someone malicious.

**Comment by psimonyi** [comment]
Indeed, as a user you have to trust whoever provides the crypto code --
and anybody whose code is running on that page for any other reason,
too. I think hard-to-guess URLs solve the privacy problem for most
practical purposes.

**Comment by bheesham** [comment]
We should have different options for "private" pastes.
- Encrypted
- Captcha
- Others?

**Comment by psimonyi** [comment]
@bheesham, what use cases are you thinking of? Let's list why someone would want especially private code; we have to know what the threat is before we can defend against it. (Remember that right now, you have to know the URL to get at the code anyway; it's not public.)

**Comment by spratt** [comment]
i'm marking this as an extension since I think this requires SSL which we won't be getting in the near future.

I see the academic use case for encrypted pages.  A TA wanting to give private feedback on student code probably doesn't want to let other see the feedback without the express permission of the student (in which case, the student can give that person the password).

**Comment by psimonyi** [comment]
I think this is a really poorly specified issue. If it's "we should
serve pages over SSL", then sure. But if it's some sort of "we should
have passwords" then I'm unconvinced. I guess we don't have to sort this
out to call it a potential extension, though.

**Comment by spratt** [comment]
I interpret it as we need to provide private pages.  See my write-up in the Extensions.md of the docs repo.
